lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp

```diff
@@ -0,0 +1,250 @@
+#include "SwiftUnsafeTypes.h"
+
+#include "lldb/DataFormatters/TypeSynthetic.h"
+#include "lldb/Symbol/SwiftASTContext.h"
+#include "lldb/Target/SwiftLanguageRuntime.h"
+
+#include <utility>
+
+using namespace lldb;
+using namespace lldb_private;
+
+class SwiftUnsafeBufferPointer {
+public:
+  SwiftUnsafeBufferPointer(ValueObject &valobj);
+  size_t GetCount() const { return m_count; }
+  addr_t GetStartAddress() const { return m_start_addr; }
+  CompilerType GetElementType() const { return m_elem_type; }
+  bool Update();
+
+private:
+  ValueObject &m_valobj;
+  size_t m_count;
+  addr_t m_start_addr;
+  CompilerType m_elem_type;
+};
+
+SwiftUnsafeBufferPointer::SwiftUnsafeBufferPointer(ValueObject &valobj)
+    : m_valobj(*valobj.GetNonSyntheticValue().get()) {}
+
+bool SwiftUnsafeBufferPointer::Update() {
+  if (!m_valobj.GetNumChildren())
+    return false;
+
+  // Here is the layout of Swift's Unsafe[Mutable]BufferPointer.
+  //
+  //  ▿ UnsafeBufferPointer
+  //    ▿ _position : Optional<UnsafePointer<Int>>
+  //      ▿ some : UnsafePointer<Int>
+  //        - pointerValue : Int
+  //    - count : Int
+  //
+  // The structure has 2 children:
+  //  1. The buffer `count` child stored as a Swift `Int` type. This entry is a
+  // "value-providing synthetic children", so lldb need to access to its
+  // children in order to get the  actual value.
+  //  2. An Optional UnsafePointer to the buffer start address. To access the
+  // pointer address, lldb unfolds every value object child until reaching
+  // `pointerValue`.
+
+  static ConstString g_count("count");
+  ValueObjectSP count_value_sp(m_valobj.GetChildMemberWithName(g_count, true));
+  if (!count_value_sp)
+    return false;
+
+  ValueObjectSP value_provided_child_sp = nullptr;
+
+  // Implement Swift's 'value-providing synthetic children' workaround.
+  // Depending on whether the value object type is a primitive or a structure,
+  // lldb should prioritize the synthetic value children.
+  // If it has no synthetic children then fallback to non synthetic children.
+  ValueObjectSP synthetic = count_value_sp->GetSyntheticValue();
+  if (synthetic)
+    value_provided_child_sp = synthetic->GetChildAtIndex(0, true);
+  if (!value_provided_child_sp)
+    value_provided_child_sp = count_value_sp->GetChildAtIndex(0, true);
+  // If neither child exists, fail.
+  if (!value_provided_child_sp)
+    return false;
+
+  m_count = value_provided_child_sp->GetValueAsUnsigned(UINT64_MAX);
```

Comment:
I don't know under which conditions this can fail, but the API makes it look like it can. I would add a defensive check for UINT64_MAX after this.